datumaro/cli/contexts/project/__init__.py

```diff
@@ -437,17 +437,17 @@ def build_transform_parser(parser_ctor=argparse.ArgumentParser):
         saved if '--apply' is enabled.|n
         |n
         Examples:|n
-        - Convert instance polygons to masks:|n
+        - Convert instance polygons to masks:|n |n
         |s|s%(prog)s -t polygons_to_masks|n
         |n
         - Rename dataset items by a regular expression|n
-        |s|s- Replace 'pattern' with 'replacement'|n|n
-        |s|s%(prog)s -t rename -- -e '|pattern|replacement|'|n
+        |s|s- Replace 'pattern' with 'replacement'|n |n
```

Comment:
I see this `|n |n` pattern in many places, what it is supposed for? It will add an extra empty line in CLI output.

Reply:
Yes, 
This help is currently displayed like this:
![](https://user-images.githubusercontent.com/54434686/155122485-6765ec5a-50dd-460f-b583-79a40de02822.png)
And it will be displayed like this:
![](https://user-images.githubusercontent.com/54434686/155122449-6792be90-0fdc-4527-bf31-0744dcda3bd1.png)

Reply:
It's hard to say which one is better, but anyway. Could you maintain the same formatting in code then and add a newline? 

Also, maybe we should just use a single level list here?

Reply:
<details>
<summary>Now the output of the help will look like this:</summary>

```console
usage: datum transform [-h] -t TRANSFORM [-o DST_DIR] [--overwrite]
                       [-p PROJECT_DIR] [--stage STAGE] [--apply APPLY]
                       [target] ...

Applies a batch operation to a dataset and produces a new dataset.

By default, datasets are updated in-place. The '-o/--output-dir' option can be
used to specify another output directory. When updating in-place, use the '--
overwrite' parameter (in-place updates fail by default to prevent data loss),
unless a project target is modified.

Builtin transforms: anns_to_labels, bbox_values_decrement, boxes_to_masks,
crop_covered_segments, id_from_image_name, label_random_sampler, map_subsets,
masks_to_polygons, merge_instance_segments, ndr, polygons_to_masks,
project_labels, random_sampler, random_split, reindex, remap_labels,
remove_annotations, remove_attributes, remove_items, rename, resize,
shapes_to_boxes, split

This command has the following invocation syntax: - datum transform <target
dataset revpath>

<revpath> - either a dataset path or a revision path. The full syntax is:
- Dataset paths:
  - <dataset path>[ :<format> ]
- Revision paths:
  - <project path> [ @<rev> ] [ :<target> ]
  - <rev> [ :<target> ]
  - <target>

The current project (-p/--project) is also used as a context for plugins, so
it can be useful for dataset paths having custom formats. When not specified,
the current project's working tree is used.

The command can be applied to a dataset or a project build target, a stage or
the combined 'project' target, in which case all the targets will be affected.
A build tree stage will be recorded if '--stage' is enabled, and the resulting
dataset(-s) will be saved if '--apply' is enabled.

Examples:
- Convert instance polygons to masks:

  datum transform -t polygons_to_masks

- Rename dataset items by a regular expression
  - Replace 'pattern' with 'replacement'

    datum transform -t rename -- -e '|pattern|replacement|'

  - Remove 'frame_' from item ids

    datum transform -t rename -- -e '|frame_(\d+)|\1|'

- Split a dataset randomly:

  datum transform -t random_split --overwrite path/to/dataset:voc

positional arguments:
  target                Target dataset revpath (default: project)
  extra_args            Additional arguments for transformation (pass '-- -h'
                        for help). Must be specified after the main command
                        arguments and after the '--' separator

optional arguments:
  -h, --help            show this help message and exit
  -t TRANSFORM, --transform TRANSFORM
                        Transform to apply to the dataset
  -o DST_DIR, --output-dir DST_DIR
                        Output directory. Can be omitted for main project
                        targets (i.e. data sources and the 'project' target,
                        but not intermediate stages) and dataset targets. If
                        not specified, the results will be saved inplace.
  --overwrite           Overwrite existing files in the save directory
  -p PROJECT_DIR, --project PROJECT_DIR
                        Directory of the project to operate on (default:
                        current dir)
  --stage STAGE         Include this action as a project build step. If true,
                        this operation will be saved in the project build
                        tree, allowing to reproduce the resulting dataset
                        later. Applicable only to main project targets (i.e.
                        data sources and the 'project' target, but not
                        intermediate stages) (default: True)
  --apply APPLY         Run this command immediately. If disabled, only the
                        build tree stage will be written (default: True)
--------------------------------------------------------------------- 
usage: rename [-h] [-e REGEX]

Renames items in the dataset. Supports regular expressions. The first
character in the expression is a delimiter for the pattern and replacement
parts. Replacement part can also contain `str.format` replacement fields with
the `item` (of type `DatasetItem`) object available.

Examples:
  - Replace 'pattern' with 'replacement':

    rename -e '|pattern|replacement|'

  - Remove 'frame_' from item ids:

    rename -e '|^frame_||'

  - Rename by regex:

    rename -e '|frame_(\d+)_extra|{item.subset}_id_\1|'

optional arguments:
  -h, --help            show this help message and exit
  -e REGEX, --regex REGEX
                        Regex for renaming in the form
                        '<sep><search><sep><replacement><sep>'
--------------------------------------------------------------------- 
usage: remap_labels [-h] [-l MAPPING] [--default {keep,delete}]

Changes labels in the dataset.

A label can be:
  - renamed (and joined with existing) -
    when '--label <old_name>:<new_name>' is specified
  - deleted - when '--label <name>:' is specified, or default action
    is 'delete' and the label is not mentioned in the list.
    When a label is deleted, all the associated annotations are removed
  - kept unchanged - when specified '--label <name>:<name>'
    or default action is 'keep' and the label is not mentioned in the list.

Annotations with no label are managed by the default action policy.

Examples:

  - Remove the 'person' label (and corresponding annotations):

    remap_labels -l person: --default keep

  - Rename 'person' to 'pedestrian' and 'human' to 'pedestrian', join:

    remap_labels -l person:pedestrian -l human:pedestrian --default keep

  - Rename 'person' to 'car' and 'cat' to 'dog', keep 'bus', remove others:

    remap_labels -l person:car -l bus:bus -l cat:dog --default delete

optional arguments:
  -h, --help            show this help message and exit
  -l MAPPING, --label MAPPING
                        Label in the form of: '<src>:<dst>' (repeatable)
  --default {keep,delete}
                        Action for unspecified labels (default: keep)
--------------------------------------------------------------------- 
usage: project_labels [-h] [-l DST_LABELS]

Changes the order of labels in the dataset from the existing to the desired
one, removes unknown labels and adds new labels. Updates or removes the
corresponding annotations.

Labels are matched by names (case dependent). Parent labels are only kept if
they are present in the resulting set of labels. If new labels are added, and
the dataset has mask colors defined, new labels will obtain generated colors.

Useful for merging similar datasets, whose labels need to be aligned.

Examples:
  - Align the source dataset labels to [person, cat, dog]:

    project_labels -l person -l cat -l dog

optional arguments:
  -h, --help            show this help message and exit
  -l DST_LABELS, --label DST_LABELS
                        Label name (repeatable, ordered)
--------------------------------------------------------------------- 
usage: resize [-h] [-dw WIDTH] [-dh HEIGHT]

Resizes images and annotations in the dataset to the specified size. Supports
upscaling, downscaling and mixed variants.

Examples:
- Resize all images to 256x256 size

  resize -dw 256 -dh 256

optional arguments:
  -h, --help            show this help message and exit
  -dw WIDTH, --width WIDTH
                        Destination image width
  -dh HEIGHT, --height HEIGHT
                        Destination image height
--------------------------------------------------------------------- 
usage: remove_items [-h] --id IDS

Allows to remove specific dataset items from dataset by their ids.

Can be useful to clean the dataset from broken or unnecessary samples.

Examples:
- Remove specific items from the dataset

  remove_items --id 'image1:train' --id 'image2:test'

optional arguments:
  -h, --help  show this help message and exit
  --id IDS    Item id to remove. Id is 'name:subset' pair (repeatable)
--------------------------------------------------------------------- 
usage: remove_annotations [-h] [--id IDS]

Allows to remove annotations on specific dataset items.

Can be useful to clean the dataset from broken or unnecessary annotations.

Examples:
- Remove annotations from specific items in the dataset

  remove_annotations --id 'image1:train' --id 'image2:test'

optional arguments:
  -h, --help  show this help message and exit
  --id IDS    Image id to clean from annotations. Id is 'name:subset' pair. If
              not specified, removes all annotations (repeatable)
--------------------------------------------------------------------- 
usage: remove_attributes [-h] [--id IDS] [-a ATTRIBUTES]

Allows to remove item and annotation attributes in a dataset.

Can be useful to clean the dataset from broken or unnecessary attributes.

Examples:
- Remove the `is_crowd` attribute from dataset

  remove_attributes --attr 'is_crowd'

- Remove the `occluded` attribute from annotations of
  the `2010_001705` item in the `train` subset

  remove_attributes --id '2010_001705:train' --attr 'occluded'

optional arguments:
  -h, --help            show this help message and exit
  --id IDS              Image id to clean from annotations. Id is
                        'name:subset' pair. If not specified, affects all
                        images and annotations (repeatable)
  -a ATTRIBUTES, --attr ATTRIBUTES
                        Attribute name to be removed. If not specified,
                        removes all attributes (repeatable)
--------------------------------------------------------------------- 
usage: ndr [-h] [-w WORKING_SUBSET] [-d DUPLICATED_SUBSET] [-a {gradient}]
           [-k NUM_CUT] [-e {random,similarity}] [-u {uniform,inverse}]
           [-s SEED]

Removes near-duplicated images in subset

Remove duplicated images from a dataset. Keep at most `-k/--num_cut` resulting
images.

Available oversampling policies (the `-e` parameter):
- `random` - sample from removed data randomly
- `similarity` - sample from removed data with ascending similarity score

Available undersampling policies (the `-u` parameter):
- `uniform` - sample data with uniform distribution
- `inverse` - sample data with reciprocal of the number of number of items
with the same similarity

Example: apply NDR, return no more than 100 images

  ndr
    --working_subset train
    --algorithm gradient
    --num_cut 100
    --over_sample random
    --under_sample uniform

optional arguments:
  -h, --help            show this help message and exit
  -w WORKING_SUBSET, --working_subset WORKING_SUBSET
                        Name of the subset to operate (default: None)
  -d DUPLICATED_SUBSET, --duplicated_subset DUPLICATED_SUBSET
                        Name of the subset for the removed data after NDR runs
                        (default: duplicated)
  -a {gradient}, --algorithm {gradient}
                        Name of the algorithm to use (default: gradient)
  -k NUM_CUT, --num_cut NUM_CUT
                        Maximum output dataset size
  -e {random,similarity}, --over_sample {random,similarity}
                        The policy to use when num_cut is bigger than result
                        length (default: random)
  -u {uniform,inverse}, --under_sample {uniform,inverse}
                        The policy to use when num_cut is smaller than result
                        length (default: uniform)
  -s SEED, --seed SEED  Random seed
--------------------------------------------------------------------- 
usage: split [-h] [-t {classification,detection,segmentation,reid}]
             [-s SPLITS] [--query QUERY] [--attr ATTR_FOR_ID] [--seed SEED]

- classification split
    Splits dataset into subsets(train/val/test) in class-wise manner.
    Splits dataset images in the specified ratio, keeping the initial class
    distribution.

- detection & segmentation split
    Each image can have multiple object annotations -
    (bbox, mask, polygon). Since an image shouldn't be included
    in multiple subsets at the same time, and image annotations
    shouldn't be split, in general, dataset annotations are unlikely
    to be split exactly in the specified ratio.
    This split tries to split dataset images as close as possible
    to the specified ratio, keeping the initial class distribution.

- reidentification split
    In this task, the test set should consist of images of unseen
    people or objects during the training phase.
    This function splits a dataset in the following way:

  1. Splits the dataset into 'train + val' and 'test' sets
     based on person or object ID.
  2. Splits 'test' set into 'test-gallery' and 'test-query' sets
     in class-wise manner.
  3. Splits the 'train + val' set into 'train' and 'val' sets
     in the same way.

The final subsets would be
'train', 'val', 'test-gallery' and 'test-query'.

Notes:
  - Each image is expected to have only one Annotation. Unlabeled or
    multi-labeled images will be split into subsets randomly.
  - If Labels also have attributes, also splits by attribute values.
  - If there is not enough images in some class or attributes group,
    the split ratio can't be guaranteed.
    In reidentification task,
  - Object ID can be described by Label, or by attribute (--attr parameter)
  - The splits of the test set are controlled by '--query' parameter
    Gallery ratio would be 1.0 - query.

Example:

  split -t classification --subset train:.5 --subset val:.2 --subset test:.3
  split -t detection --subset train:.5 --subset val:.2 --subset test:.3
  split -t segmentation --subset train:.5 --subset val:.2 --subset test:.3
  split -t reid --subset train:.5 --subset val:.2 --subset test:.3 --query .5

Example: use 'person_id' attribute for splitting

  split --attr person_id

optional arguments:
  -h, --help            show this help message and exit
  -t {classification,detection,segmentation,reid}, --task {classification,detection,segmentation,reid}
                        (one of classification, detection, segmentation, reid;
                        default: classification)
  -s SPLITS, --subset SPLITS
                        Subsets in the form: '<subset>:<ratio>' (repeatable,
                        default: {'train': 0.5, 'val': 0.2, 'test': 0.3})
  --query QUERY         Query ratio in the test set (default: 0.500)
  --attr ATTR_FOR_ID    Attribute name representing the ID (default: use
                        label)
  --seed SEED           Random seed
--------------------------------------------------------------------- 
usage: random_sampler [-h] -k COUNT [-s SUBSET] [--seed SEED]

Sampler that keeps no more than required number of items in the dataset.

Notes:
  - Items are selected uniformly
  - Requesting a sample larger than the number of all images will
    return all images

Example: select subset of 20 images randomly

  random_sampler -k 20

Example: select subset of 20 images, modify only 'train' subset

  random_sampler -k 20 -s train

optional arguments:
  -h, --help            show this help message and exit
  -k COUNT, --count COUNT
                        Maximum number of items to sample
  -s SUBSET, --subset SUBSET
                        Limit changes to this subset (default: affect all
                        dataset)
  --seed SEED           Initial value for random number generator
--------------------------------------------------------------------- 
usage: label_random_sampler [-h] -k COUNT [-l LABEL_COUNTS] [--seed SEED]

Sampler that keeps at least the required number of annotations of each class
in the dataset for each subset separately.

Consider using the "stats" command to get class distribution in the dataset.

Notes:
  - Items can contain annotations of several selected classes
    (e.g. 3 bounding boxes per image). The number of annotations in the
    resulting dataset varies between max(class counts) and sum(class counts)
  - If the input dataset does not has enough class annotations, the result
    will contain only what is available
  - Items are selected uniformly
  - For reasons above, the resulting class distribution in the dataset may
    not be the same as requested
  - The resulting dataset will only keep annotations for
    classes with specified count > 0

Example: select at least 5 annotations of each class randomly

  label_random_sampler -k 5

Example: select at least 5 images with "cat" annotations and 3 "person"

  label_random_sampler -l "cat:5" -l "person:3"

optional arguments:
  -h, --help            show this help message and exit
  -k COUNT, --count COUNT
                        Minimum number of annotations of each class
  -l LABEL_COUNTS, --label LABEL_COUNTS
                        Minimum number of annotations of a specific class.
                        Overrides the `-k/--count` setting for the class. The
                        format is 'label_name:count' (repeatable)
  --seed SEED           Initial value for random number generator
```

</details>

Reply:
> Also, maybe we should just use a single level list here?

This list has several second-level items, perhaps we should leave a few levels?

```console

usage: datum transform [-h] -t TRANSFORM [-o DST_DIR] [--overwrite]
                       [-p PROJECT_DIR] [--stage STAGE] [--apply APPLY]
                       [target] ...

Applies a batch operation to a dataset and produces a new dataset.

By default, datasets are updated in-place. The '-o/--output-dir' option can be
used to specify another output directory. When updating in-place, use the '--
overwrite' parameter (in-place updates fail by default to prevent data loss),
unless a project target is modified.

Builtin transforms: anns_to_labels, bbox_values_decrement, boxes_to_masks,
crop_covered_segments, id_from_image_name, label_random_sampler, map_subsets,
masks_to_polygons, merge_instance_segments, ndr, polygons_to_masks,
project_labels, random_sampler, random_split, reindex, remap_labels,
remove_annotations, remove_attributes, remove_items, rename, resize,
shapes_to_boxes, split

This command has the following invocation syntax: - datum transform <target
dataset revpath>

<revpath> - either a dataset path or a revision path. The full syntax is:
- Dataset paths:
  - <dataset path>[ :<format> ]
- Revision paths:
  - <project path> [ @<rev> ] [ :<target> ]
  - <rev> [ :<target> ]
  - <target>

The current project (-p/--project) is also used as a context for plugins, so
it can be useful for dataset paths having custom formats. When not specified,
the current project's working tree is used.

The command can be applied to a dataset or a project build target, a stage or
the combined 'project' target, in which case all the targets will be affected.
A build tree stage will be recorded if '--stage' is enabled, and the resulting
dataset(-s) will be saved if '--apply' is enabled.

Examples:
- Convert instance polygons to masks:

  datum transform -t polygons_to_masks

- Rename dataset items by a regular expression:

  - Replace 'pattern' with 'replacement':

    datum transform -t rename -- -e '|pattern|replacement|'

  - Remove 'frame_' from item ids:

    datum transform -t rename -- -e '|frame_(\d+)|\1|'

- Split a dataset randomly:

  datum transform -t random_split --overwrite path/to/dataset:voc

positional arguments:
  target                Target dataset revpath (default: project)
  extra_args            Additional arguments for transformation (pass '-- -h'
                        for help). Must be specified after the main command
                        arguments and after the '--' separator

optional arguments:
  -h, --help            show this help message and exit
  -t TRANSFORM, --transform TRANSFORM
                        Transform to apply to the dataset
  -o DST_DIR, --output-dir DST_DIR
                        Output directory. Can be omitted for main project
                        targets (i.e. data sources and the 'project' target,
                        but not intermediate stages) and dataset targets. If
                        not specified, the results will be saved inplace.
  --overwrite           Overwrite existing files in the save directory
  -p PROJECT_DIR, --project PROJECT_DIR
                        Directory of the project to operate on (default:
                        current dir)
  --stage STAGE         Include this action as a project build step. If true,
                        this operation will be saved in the project build
                        tree, allowing to reproduce the resulting dataset
                        later. Applicable only to main project targets (i.e.
                        data sources and the 'project' target, but not
                        intermediate stages) (default: True)
  --apply APPLY         Run this command immediately. If disabled, only the
                        build tree stage will be written (default: True)

```